File: Alcoholtracker/Views/Components/BottleGraphic.swift
import SwiftUI

// MARK: - BottleGraphic
// Stylised bottle silhouette showing start level (dashed line) and current level (filled).

struct BottleGraphic: View {

    let startLevel: Double   // 0.0 = empty, 1.0 = full
    let currentLevel: Double

    var body: some View {
        GeometryReader { geo in
            let h = geo.size.height
            let w = geo.size.width
            
            // 1.0 (100%) maps to 76% of physical height (starts below the neck)
            let maxFillRatio: Double = 0.76
            let visualStartLevel = startLevel * maxFillRatio
            let visualCurrentLevel = currentLevel * maxFillRatio

            ZStack(alignment: .bottom) {

                // Consumed region (start -> current), lighter accent tint
                BottleShape()
                    .fill(Color.appAccent.opacity(0.25))
                    .mask(
                        VStack(spacing: 0) {
                            Spacer()
                                .frame(height: h * (1 - visualStartLevel))
                            Rectangle()
                                .frame(height: h * max(0, visualStartLevel - visualCurrentLevel))
                            Spacer()
                                .frame(height: h * visualCurrentLevel)
                        }
                    )

                // Remaining liquid
                BottleShape()
                    .fill(Color.appAccent.opacity(0.55))
                    .mask(
                        VStack(spacing: 0) {
                            Spacer()
                            Rectangle()
                                .frame(height: h * visualCurrentLevel)
                        }
                    )

                // Bottle outline
                BottleShape()
                    .stroke(Color.appBorder, lineWidth: 1.5)

                // Start-level marker line
                let markerY = h * (1 - visualStartLevel)
                Path { path in
                    path.move(to: CGPoint(x: 0, y: markerY))
                    path.addLine(to: CGPoint(x: w, y: markerY))
                }
                .stroke(Color.appAccent, style: StrokeStyle(lineWidth: 1.5, dash: [4, 3]))

                // "Start" label
                if startLevel < 0.95 {
                    Text("Start")
                        .font(.system(size: 8, weight: .medium))
                        .foregroundStyle(Color.appAccent)
                        .position(x: w * 0.75, y: markerY - 8)
                }
            }
        }
    }
}

// MARK: - BottleShape

struct BottleShape: Shape {
    func path(in rect: CGRect) -> Path {
        var p = Path()
        let w = rect.width
        let h = rect.height
        let cx = w / 2
        let neckW = w * 0.28
        let bodyR: CGFloat = 8
        let neckTop: CGFloat = h * 0.04
        let neckBot: CGFloat = h * 0.24
        let shoulder: CGFloat = h * 0.32

        p.move(to: CGPoint(x: cx - neckW / 2, y: neckTop))
        p.addLine(to: CGPoint(x: cx - neckW / 2, y: neckBot))
        p.addCurve(
            to: CGPoint(x: bodyR, y: shoulder + 18),
            control1: CGPoint(x: cx - neckW / 2, y: shoulder),
            control2: CGPoint(x: bodyR, y: shoulder)
        )
        p.addLine(to: CGPoint(x: bodyR, y: h - bodyR))
        p.addArc(center: CGPoint(x: bodyR * 2, y: h - bodyR),
                 radius: bodyR, startAngle: .degrees(180), endAngle: .degrees(90), clockwise: true)
        p.addLine(to: CGPoint(x: w - bodyR * 2, y: h))
        p.addArc(center: CGPoint(x: w - bodyR * 2, y: h - bodyR),
                 radius: bodyR, startAngle: .degrees(90), endAngle: .degrees(0), clockwise: true)
        p.addLine(to: CGPoint(x: w - bodyR, y: shoulder + 18))
        p.addCurve(
            to: CGPoint(x: cx + neckW / 2, y: neckBot),
            control1: CGPoint(x: w - bodyR, y: shoulder),
            control2: CGPoint(x: cx + neckW / 2, y: shoulder)
        )
        p.addLine(to: CGPoint(x: cx + neckW / 2, y: neckTop))
        p.closeSubpath()
        return p
    }
}

File: android/app/src/main/kotlin/de/tipau/promille/ui/components/BottleGraphic.kt
package de.tipau.promille.ui.components
import androidx.compose.material3.Icon
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxWithConstraints
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.offset
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Rect
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.PathEffect
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.graphics.drawscope.clipPath
import androidx.compose.ui.graphics.drawscope.clipRect
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import de.tipau.promille.AppColors

/**
 * Port of BottleGraphic.swift.
 * Stylised bottle silhouette showing start level (dashed line) and current level (filled).
 */
@Composable
fun BottleGraphic(
    startLevel: Double,   // 0.0 = empty, 1.0 = full
    currentLevel: Double,
    modifier: Modifier = Modifier
) {
    BoxWithConstraints(modifier = modifier) {
        val widthDp = maxWidth
        val heightDp = maxHeight

        val accentColor = AppColors.accent
        val borderColor = AppColors.border

        Canvas(modifier = Modifier.fillMaxSize()) {
            val w = size.width
            val h = size.height

            // 1.0 (100%) maps to 76% of physical height (starts below the neck)
            val maxFillRatio = 0.76f
            val visualStartLevel = (startLevel.toFloat() * maxFillRatio).coerceIn(0f, maxFillRatio)
            val visualCurrentLevel = (currentLevel.toFloat() * maxFillRatio).coerceIn(0f, maxFillRatio)

            val bottlePath = createBottlePath(w, h)

            // 1. Consumed region (start -> current), lighter accent tint (0.25 opacity)
            if (visualStartLevel > visualCurrentLevel) {
                clipPath(bottlePath) {
                    val topY = h * (1f - visualStartLevel)
                    val bottomY = h * (1f - visualCurrentLevel)
                    clipRect(
                        left = 0f,
                        top = topY,
                        right = w,
                        bottom = bottomY
                    ) {
                        drawRect(color = accentColor.copy(alpha = 0.25f))
                    }
                }
            }

            // 2. Remaining liquid (current level down to bottom), darker accent tint (0.55 opacity)
            if (visualCurrentLevel > 0f) {
                clipPath(bottlePath) {
                    val topY = h * (1f - visualCurrentLevel)
                    clipRect(
                        left = 0f,
                        top = topY,
                        right = w,
                        bottom = h
                    ) {
                        drawRect(color = accentColor.copy(alpha = 0.55f))
                    }
                }
            }

            // 3. Bottle outline
            drawPath(
                path = bottlePath,
                color = borderColor,
                style = Stroke(width = 1.5.dp.toPx())
            )

            // 4. Start-level marker line (dashed)
            val markerY = h * (1f - visualStartLevel)
            val dashEffect = PathEffect.dashPathEffect(floatArrayOf(4.dp.toPx(), 3.dp.toPx()), 0f)
            drawLine(
                color = accentColor,
                start = Offset(0f, markerY),
                end = Offset(w, markerY),
                strokeWidth = 1.5.dp.toPx(),
                pathEffect = dashEffect
            )
        }

        // 5. "Start" label if startLevel < 0.95
        if (startLevel < 0.95) {
            val maxFillRatio = 0.76f
            val visualStartLevel = (startLevel.toFloat() * maxFillRatio).coerceIn(0f, maxFillRatio)
            val markerYDp = heightDp * (1f - visualStartLevel) - 10.dp
            val labelXDp = widthDp * 0.65f

            Box(
                modifier = Modifier.fillMaxSize()
            ) {
                Box(
                    modifier = Modifier.offset(x = labelXDp, y = markerYDp)
                ) {
                    Text(
                        text = "Start",
                        color = accentColor,
                        fontSize = 8.sp,
                        fontWeight = FontWeight.Medium
                    )
                }
            }
        }
    }
}

/**
 * Builds the geometric bottle outline matching BottleShape in BottleGraphic.swift.
 */
private fun createBottlePath(w: Float, h: Float): Path {
    val path = Path()
    val cx = w / 2f
    val neckW = w * 0.28f
    val bodyR = 8f
    val neckTop = h * 0.04f
    val neckBot = h * 0.24f
    val shoulder = h * 0.32f

    path.moveTo(cx - neckW / 2f, neckTop)
    path.lineTo(cx - neckW / 2f, neckBot)
    path.cubicTo(
        cx - neckW / 2f, shoulder,
        bodyR, shoulder,
        bodyR, shoulder + 18f
    )
    path.lineTo(bodyR, h - bodyR)
    path.arcTo(
        rect = Rect(bodyR, h - 2f * bodyR, 3f * bodyR, h),
        startAngleDegrees = 180f,
        sweepAngleDegrees = -90f,
        forceMoveTo = false
    )
    path.lineTo(w - 2f * bodyR, h)
    path.arcTo(
        rect = Rect(w - 3f * bodyR, h - 2f * bodyR, w - bodyR, h),
        startAngleDegrees = 90f,
        sweepAngleDegrees = -90f,
        forceMoveTo = false
    )
    path.lineTo(w - bodyR, shoulder + 18f)
    path.cubicTo(
        w - bodyR, shoulder,
        cx + neckW / 2f, shoulder,
        cx + neckW / 2f, neckBot
    )
    path.lineTo(cx + neckW / 2f, neckTop)
    path.close()
    return path
}
